R/virtualenv.R

```diff
@@ -683,23 +683,29 @@ as_version_constraint_checkers <- function(version) {
   version <- unlist(strsplit(version, ",", fixed = TRUE))
 
   # given string like ">=3.8", match two groups, on ">=" and "3.8"
-  pattern <- "^([><=!]{0,2})\\s*([0-9.a-zA-Z]*)"
+  pattern <- "^([><=!]{0,2})\\s*([0-9.a-zA-Z*]*)"
 
   op <- sub(pattern, "\\1",  version)
   op[op == ""] <- "=="
 
   ver_string <- sub(pattern, "\\2",  version)
+
+  ver_string <- sub('\\.\\*$', '', ver_string)
+
   ver <- numeric_version(ver_string, strict = FALSE)
 
-  .mapply(function(op, ver, ver_string) {
+  .mapply(function(op, ver, ver_string, version) {
     op <- as.symbol(op)
     force(ver)
     force(ver_string)
-
+    force(version)
     # return a "checker" function that takes a vector of versions and returns
     # a logical vector of if the version satisfies the constraint.
     rlang::zap_srcref(eval(bquote(function(x) {
-      op <- .(op)
+      op <- try(.(op), silent = TRUE)
+      if (inherits(op, "try-error")) {
+        stop("Version `", version, "` is not valid.")
```

Comment:
I'm not sure if it's safe to throw an error from the checker here, since this function is used in a few contexts. Also, is it throwing too early? Are we potentially throwing an error on a valid input like `3.13-dev` or `3.14.0rc1` or `3.13.6t`, i.e., strings that don't parse as semantic versions but can be matched exactally?

Reply:
I'm only returning a more specific message if the `.(op)` call fails, that's all. With out this code calling, for example: `virtualenv_starter("==3.12.-")` returns this in the current version:

`Error in check(starters$version) : object '==-' not found`

This change simply makes it clear that the version itself is what's actually problematic: 

`Error in check(starters$version) : Version `==3.12.-` is not valid.`

In other words, I'm not adding an extra check, just a more helpful error message

Reply:
Btw, it looks like `3.13-dev` fails in the current version, is that to be addressed too?

```
> reticulate::virtualenv_starter("3.13-dev")
Error in check(starters$version) : object '-dev' not found
```

It also fails on this new version, but with a better message:

```
> virtualenv_starter("3.13-dev")
Error in check(starters$version) : Version `3.13-dev` is not valid.
```